q2_demux/tests/test_filter.py

```
@@ -0,0 +1,183 @@
+        dir_files = os.listdir(str(test))
+        dir_ids = [(lambda x: dir_files[x].split('_', 1)[0])(x) for
+                   x in range(len(dir_files))]
+        meta_ids = meta.get_ids(where=where)
+
+        self.assertEqual(meta_ids, set(dir_ids))
+
+    def _single_exclude(self, meta, where):
+        test = filter.filter_samples(self.sample_single, meta,
+                                     where, True)
+        dir_files = os.listdir(str(test))
+        dir_ids = [(lambda x: dir_files[x].split('_', 1)[0])(x) for
+                   x in range(len(dir_files))]
+        meta_ids = meta.get_ids(where=where)
+        exp_ids = set(self.manifest_single.index) - meta_ids
+
+        self.assertEqual(exp_ids, set(dir_ids))
+
+    def test_filter_single(self):
+        metas = [self.meta_single_all, self.meta_single_subset]
+        wheres = [None, "Study='A'", "Study='B'", "Study='A' OR Study='B'",
+                  "Study='A' AND Study='B'"]
+        for meta in metas:
+            for where in wheres:
+                with self.subTest(meta=meta, where=where):
+                    self._single_no_exclude(meta, where)
+
+    def test_filter_single_exclude(self):
+        metas = [self.meta_single_all, self.meta_single_subset]
+        wheres = [None, "Study='A'", "Study='B'", "Study='A' OR Study='B'",
+                  "Study='A' AND Study='B'"]
+        for meta in metas:
+            for where in wheres:
+                with self.subTest(meta=meta, where=where):
+                    self._single_exclude(meta, where)
+
+    def test_filter_single_none_exclude(self):
+        wheres = [None, "Study='A'", "Study='B'", "Study='A' OR Study='B'",
+                  "Study='A' AND Study='B'"]
+        for where in wheres:
+            with self.subTest(where=where):
+                self._single_exclude(self.meta_single_none, where)
+
+    def test_filter_single_none_no_exclude_no_exception(self):
+        wheres = ["Study='B'", "Study='A' AND Study='B'"]
```

Thinking about it, I think this probably should raise an exception, as users are very likely to generate meaningless queries (as you are testing here).